File: src/lib/api/claudeAppendInstructions.ts
import { invoke } from "@tauri-apps/api/core";

export type ClaudeAppendInstructionsFileState =
  | "valid"
  | "missing"
  | "notFile"
  | "unreadable"
  | "empty"
  | "invalid";

export interface ClaudeAppendInstructionsConfig {
  files: string[];
  activeFile?: string | null;
}

export interface ClaudeAppendInstructionsFileStatus {
  configuredPath: string;
  resolvedPath: string;
  state: ClaudeAppendInstructionsFileState;
  exists: boolean;
  isFile: boolean;
  isSymlink: boolean;
  readable: boolean;
  sizeBytes: number | null;
  modifiedAt: number | null;
  sha256: string | null;
  error: string | null;
}

export const claudeAppendInstructionsApi = {
  async getConfig(): Promise<ClaudeAppendInstructionsConfig> {
    return await invoke("get_claude_append_instructions_config");
  },

  async setConfig(
    config: ClaudeAppendInstructionsConfig,
  ): Promise<ClaudeAppendInstructionsConfig> {
    return await invoke("set_claude_append_instructions_config", { config });
  },

  async inspect(
    configuredPath: string,
  ): Promise<ClaudeAppendInstructionsFileStatus> {
    return await invoke("inspect_claude_append_instructions_file", {
      configuredPath,
    });
  },

  async read(configuredPath: string): Promise<string | null> {
    return await invoke("read_claude_append_instructions_file", {
      configuredPath,
    });
  },

  async write(
    configuredPath: string,
    content: string,
  ): Promise<ClaudeAppendInstructionsFileStatus> {
    return await invoke("write_claude_append_instructions_file", {
      configuredPath,
      content,
    });
  },

  async remove(configuredPath: string): Promise<boolean> {
    return await invoke("delete_claude_append_instructions_file", {
      configuredPath,
    });
  },
};

File: src-tauri/src/commands/claude_append_instructions.rs
use tauri::State;

use crate::claude_append_instructions::{
    self, ClaudeAppendInstructionsConfig, ClaudeAppendInstructionsFileStatus,
};
use crate::store::AppState;

#[tauri::command]
pub async fn get_claude_append_instructions_config(
    state: State<'_, AppState>,
) -> Result<ClaudeAppendInstructionsConfig, String> {
    claude_append_instructions::get_config(state.db.as_ref()).map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn set_claude_append_instructions_config(
    config: ClaudeAppendInstructionsConfig,
    state: State<'_, AppState>,
) -> Result<ClaudeAppendInstructionsConfig, String> {
    claude_append_instructions::update_config(state.db.as_ref(), config)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn inspect_claude_append_instructions_file(
    configuredPath: String,
) -> Result<ClaudeAppendInstructionsFileStatus, String> {
    tauri::async_runtime::spawn_blocking(move || {
        claude_append_instructions::inspect_file(&configuredPath)
    })
    .await
    .map_err(|error| format!("Failed to inspect Claude append instructions file: {error}"))
}

#[tauri::command]
pub async fn read_claude_append_instructions_file(
    configuredPath: String,
) -> Result<Option<String>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        claude_append_instructions::read_file(&configuredPath)
    })
    .await
    .map_err(|error| format!("Failed to read Claude append instructions file: {error}"))?
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn write_claude_append_instructions_file(
    configuredPath: String,
    content: String,
    state: State<'_, AppState>,
) -> Result<ClaudeAppendInstructionsFileStatus, String> {
    let db = state.db.clone();
    tauri::async_runtime::spawn_blocking(move || {
        claude_append_instructions::write_file(db.as_ref(), &configuredPath, &content)
    })
    .await
    .map_err(|error| format!("Failed to write Claude append instructions file: {error}"))?
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn delete_claude_append_instructions_file(
    configuredPath: String,
    state: State<'_, AppState>,
) -> Result<bool, String> {
    let db = state.db.clone();
    tauri::async_runtime::spawn_blocking(move || {
        claude_append_instructions::delete_file(db.as_ref(), &configuredPath)
    })
    .await
    .map_err(|error| format!("Failed to delete Claude append instructions file: {error}"))?
    .map_err(|error| error.to_string())
}
